cmd/gateway/nas/gateway-nas.go

```diff
@@ -0,0 +1,251 @@
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package nas
+
+import (
+	"fmt"
+	"io"
+
+	"github.com/minio/cli"
+	"github.com/minio/minio-go/pkg/policy"
+	minio "github.com/minio/minio/cmd"
+	"github.com/minio/minio/pkg/auth"
+	"github.com/minio/minio/pkg/hash"
+)
+
+const (
+	nasBackend = "nas"
+)
+
+var (
+	// Path format is not valid.
+	errNASInvalidPath = fmt.Errorf("Path to NAS storage is either empty or invalid")
+)
+
+func init() {
+	const nasGatewayTemplate = `NAME:
+  {{.HelpName}} - {{.Usage}}
+
+USAGE:
+  {{.HelpName}} {{if .VisibleFlags}}[FLAGS]{{end}} PATH
+{{if .VisibleFlags}}
+FLAGS:
+  {{range .VisibleFlags}}{{.}}
+  {{end}}{{end}}
+PATH:
+  path to NAS storage should be provided.
+
+ENVIRONMENT VARIABLES:
+  ACCESS:
+     MINIO_ACCESS_KEY: Username or access key of minimum 3 characters in length.
+     MINIO_SECRET_KEY: Password or secret key of minimum 8 characters in length.
+
+  BROWSER:
+     MINIO_BROWSER: To disable web browser access, set this value to "off".
+
+  UPDATE:
+     MINIO_UPDATE: To turn off in-place upgrades, set this value to "off".
+
+EXAMPLES:
+  1. Start minio gateway server for NAS backend.
+      $ export MINIO_ACCESS_KEY=accesskey
+      $ export MINIO_SECRET_KEY=secretkey
+      $ {{.HelpName}} PATH
+
+`
+
+	minio.RegisterGatewayCommand(cli.Command{
+		Name:               nasBackend,
+		Usage:              "Network-attached storage (NAS).",
+		Action:             nasGatewayMain,
+		CustomHelpTemplate: nasGatewayTemplate,
+		HideHelpCommand:    true,
+	})
+}
+
+// Handler for 'minio gateway nas' command line.
+func nasGatewayMain(ctx *cli.Context) {
+	// Validate gateway arguments.
+	host := ctx.Args().First()
+	if host == "" {
+		minio.ErrorIf(errNASInvalidPath, "Unable to start NAS gateway with %s", ctx.Args().First())
+		cli.ShowCommandHelpAndExit(ctx, "nas", 1)
+	}
+	// Validate gateway arguments.
+	minio.StartGateway(ctx, &NAS{host})
+}
+
+// NAS implements Gateway.
+type NAS struct {
+	host string
+}
+
+// Name implements Gateway interface.
+func (g *NAS) Name() string {
+	return nasBackend
+}
+
+// NewGatewayLayer returns nas gatewaylayer.
+func (g *NAS) NewGatewayLayer(creds auth.Credentials) (minio.ObjectLayer, error) {
+	var err error
+	newObject, err := minio.NewFSObjectLayer(g.host)
+	if err != nil {
+		return nil, err
+	}
+	return &nasObjects{
+		objAPI: newObject,
+	}, nil
+}
+
+// Production - nas gateway is not production ready.
+func (g *NAS) Production() bool {
+	return false
+}
+
+// nasObjects implements gateway for Minio and S3 compatible object storage servers.
+type nasObjects struct {
+	minio.GatewayUnsupported
+	objAPI minio.ObjectLayer
+}
```

Comment:
You can do this and remove all the other methods on nasObjects (minio.fsObjects should be exported as minio.FSObjects)
```
// nasObjects implements gateway for Minio and S3 compatible object storage servers.
type nasObjects struct {
        minio.FSObjects
}

func (l *nasObjects) IsNotificationSupported() bool {
        false
}

```